core/src/main/java/org/mule/processor/AbstractRequestResponseMessageProcessor.java

```diff
@@ -0,0 +1,150 @@
+            public void onCompletion(MuleEvent event)
+            {
+                try
+                {
+                    MuleEvent response = processResponse(new DefaultMuleEvent(event, originalReplyToHandler));
+                    if (!NonBlockingVoidMuleEvent.getInstance().equals(response))
+                    {
+                        originalReplyToHandler.processReplyTo(response, null, null);
+                    }
+                    processFinallly(event, null);
+                }
+                catch (MuleException e)
+                {
+                    onFailure(new MessagingException(event, e));
+                }
+            }
+
+            @Override
+            public void onFailure(MessagingException exception)
+            {
+                originalReplyToHandler.processExceptionReplyTo(exception.getEvent(), exception, null);
+                processFinallly(exception.getEvent(), exception);
+            }
+        }));
+
+        try
+        {
+            MuleEvent result = processNext(processRequest(event));
+            if (!(result instanceof NonBlockingVoidMuleEvent))
+            {
+                return processResponse(event);
+            }
+            else
+            {
+                return result;
+            }
+        }
+        catch (MessagingException exception)
+        {
+            processFinallly(event, exception);
+            throw exception;
+        }
+    }
+
+    private boolean isNonBlocking(MuleEvent event)
+    {
+        return event.isAllowNonBlocking() && event.getReplyToHandler() != null && next != null;
+    }
+
+    /**
+     * Processes the request phase before the next message processor is invoked.
+     *
+     * @param event event to be processed.
+     * @return result of request processing.
+     * @throws MuleException exception thrown by implementations of this method whiile performing response processing
+     */
+    protected MuleEvent processRequest(MuleEvent event) throws MuleException
+    {
+        return event;
+    }
+
+    /**
+     * Processes the response phase after the next message processor and it's response phase have been invoked
+     *
+     * @param event event to be processed.
+     * @return result of response processing.
+     * @throws MuleException exception thrown by implementations of this method whiile performing response processing
+     */
+    protected MuleEvent processResponse(MuleEvent event) throws MuleException
+    {
+        return event;
+    }
+
+    /**
+     * @param event     the result of request and response processing. Note that this includes the request and response
+     *                  processing of the rest of the Flow following this message processor too.
+     * @param exception the exception thrown during processing if any. If not exception was thrown then this parameter
+     *                  is null
+     */
+    protected void processFinallly(MuleEvent event, MessagingException exception)
```

Comment:
Why not having two different methods, one for the successful case and another one for the error?